lib/log.dart

```diff
@@ -30,3 +32,51 @@ bool debugLog(String message) {
   }());
   return true;
 }
+
+typedef ReportErrorCallback = void Function(String? message, {String? details});
+/// Display an error message in a [SnackBar].
+///
+/// This shows a [SnackBar] containing the message if [ZulipApp] is ready,
+/// otherwise logs it to the console.
+///
+/// If `message` is null, this will clear the existing [SnackBar]s if there
+/// are any.  Useful for promptly dismissing errors.
+///
+/// If `details` is non-null, the [SnackBar] will contain a button that would
+/// open a dialog containing the error details.
+// This gets set in [ZulipApp].  We need this indirection to keep `lib/log.dart`
+// from importing widget code, because the file is a dependency for the rest of
+// the app.
+ReportErrorCallback get reportErrorToUser => (message, {details}) {
+  assert(() {
+    _debugLastReportedError = details != null ? '$message\n$details' : message;
+    return true;
+  }());
+  (_reportErrorToUser ??= _defaultReportErrorToUser)(message, details: details);
+};
```

Comment:
This structure with a getter, that returns a fixed function expression, that mostly delegates to a function-typed field, that gets lazily initialized to a default value, feels a bit hard to understand.

It seems like the reason the getter is here, vs. just a field, is to inject this recording of `_debugLastReportedError`, to support `debugTakeLastReportedError`. Can we have the tests just look for the snack bar instead?

That'd make the tests more end-to-end, testing the actual user-facing behavior, which is good:
https://zulip.readthedocs.io/en/latest/testing/philosophy.html#integration-testing-or-unit-testing

And it seems like it'd simplify this code, too.

Reply:
It will indeed be a helpful refactor to simplify the entire thing to a single line.

This is mostly used for `test/model/store_test.dart` only, and we test it end-to-end in `test/widgets/app_test.dart`. I think we don't instantiate `TestZulipApp` for tests in `test/model`, but this can be easier resolved by assigning a custom `reportErrorToUserBriefly` that updates a variable like `checkNotified` does.